lib/json_validator.rb

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) 2013-2015 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of version 3 of the GNU General Public License as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE about this file by physical or electronic mail,
+# you may find current contact information at www.suse.com
+
+class JsonValidator
+  def initialize(format_version)
+    @global_schema = global_schema(format_version)
+    @scope_schemas = scope_schemas(format_version)
+  end
+
+  def validate(hash)
+    JSON::Validator.fully_validate(@global_schema, hash)
+  end
+
+  def validate_scope(hash, scope)
+    return [] if !@scope_schemas[scope]
+
+    errors = JSON::Validator.fully_validate(@scope_schemas[scope], hash).map do |error|
+      "In scope #{scope}: #{error}"
+    end
+    errors.map do |error|
+      cleanup_json_error_message(error, scope)
+    end
+  end
+
+  def global_schema(format_version = SystemDescription::CURRENT_FORMAT_VERSION)
+    JSON.parse(File.read(File.join(
+      Machinery::ROOT,
+      "schema",
+      "v#{format_version}",
+      "system-description-global.schema.json"
+    )))
+  end
+
+  def scope_schemas(format_version = SystemDescription::CURRENT_FORMAT_VERSION)
+    schema_dir = File.join(
+      Machinery::ROOT,
+      "plugins",
+      "schema",
+      "v#{format_version}",
+    )
+
+    Hash[
+      Dir["#{schema_dir}/*.schema.json"].map do |file|
+        scope = file.match(/system-description-(.*)\.schema\.json$/)[1].tr("-", "_")
+        schema = JSON.parse(File.read(file))
+
+        [scope, schema]
+      end
+    ]
+  end
+
+  private
```

Comment:
I would not have a problem with making the methods public. Then the tests don't have to use `send`.

Alternatively these methods could also be put in a separate class, something like `JsonValidationCleaner`.